### PR TITLE
add multi particle support for plugins

### PR DESCRIPTION
**Pull request by psychocoderHPC**
This pull changes many names for plugin names and plugin output files -> **explicitly mention in the next changelog**.

- change plugin prefix
- update macro documentation
- update names to activate plugins in all `*.cfg` files
- update tool `plotNumericalHeating`

**This pull request effected all `*.cfg` files**

~~Please do not merge before #802~~

@PrometheusPi please also take a look to this pull request

### CODE REVIEW

**Comment by ax3l** [comment]
Great pull, thank you!
Nice you updated `doc/TBG_macros.cfg`, too :)

File: doc/TBG_macros.cfg [hunk 1]
```diff
@@ -83,88 +83,72 @@ TBG_periodic="--periodic 1 0 1"
 # Enables moving window (sliding) in your simulation
 TBG_movingWindow="-m"
 
+################################################################################
+## Placeholder for multi data plugins:
+##  
+## placeholders must be substituted with the real data name
+##
+## SPECIES = species name e.g. e (electrons), i (ions)
```
Comment (ax3l) [inline doc/TBG_macros.cfg]: let's write `<species>` here:

```
placeholders must be substituted with the real data name (remove the <>)

<species> = species name e.g. e (electrons), i (ions)
<fields> = field names e.g. FieldE, FieldB, FieldJ
```

File: doc/TBG_macros.cfg [hunk 2]
```diff
@@ -191,10 +175,10 @@ TBG_restart="--restart"
 #   --restart-directory /path/to/simOutput/checkpoints
 
 # Connect to a live-view server (start the server in advance)
-TBG_liveViewYX="--live_e.period 1 --live_e.slicePoint 0.5 --live_e.ip 10.0.2.254 \
-                --live_e.port 2020 --live_e.axis yx"
-TBG_liveViewYZ="--live_e.period 1 --live_e.slicePoint 0.5 --live_e.ip 10.0.2.254 \
-                --live_e.port 2021 --live_e.axis yz"
+TBG_liveViewYX="--SPECIES_liveView.period 1 --SPECIES_liveView..slicePoint 0.5 --SPECIES_liveView..ip 10.0.2.254 \
+                --SPECIES_liveView..port 2020 --SPECIES_liveView..axis yx"
```
Comment (ax3l) [inline doc/TBG_macros.cfg]: typo ".."

File: doc/TBG_macros.cfg [hunk 3]
```diff
@@ -191,10 +175,10 @@ TBG_restart="--restart"
...
+TBG_liveViewYZ="--SPECIES_liveView..period 1 --SPECIES_liveView.slicePoint 0.5 --SPECIES_liveView.ip 10.0.2.254 \
```
Comment (ax3l) [inline doc/TBG_macros.cfg]: typo ".."

File: src/libPMacc/include/math/vector/compile-time/Vector.hpp [hunk 4]
```diff
@@ -97,7 +97,13 @@ struct VectorFromCT<3>
 template<typename Arg0>
 struct TypeSelector
 {
-    typedef typename Arg0::value_type type;
+    typedef Arg0 type;
```
Comment (ax3l) [inline src/libPMacc/include/math/vector/compile-time/Vector.hpp]: huh, what happened here? :)

File: src/picongpu/include/plugins/CountParticles.hpp [hunk 5]
```diff
@@ -71,15 +71,16 @@ class CountParticles : public ISimulationPlugin
...
+    CountParticles() :
+    analyzerName("CountParticles: count macro particles of a species"),
+    analyzerPrefix(ParticlesType::FrameType::getName() + std::string("_macroParticlesCount")),
+    filename(analyzerPrefix + ".dat"),
     particles(NULL),
     cellDescription(NULL),
     notifyPeriod(0),
     writeToFile(false)
     {
+
```
Comment (ax3l) [inline src/picongpu/include/plugins/CountParticles.hpp]: :) (remove me)

File: src/libPMacc/include/math/vector/compile-time/Vector.hpp (hunk 4 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/math/vector/compile-time/Vector.hpp]: magic ^^. I move this to a separate pull

Comment (psychocoderHPC) [inline src/libPMacc/include/math/vector/compile-time/Vector.hpp]: moved to #802

Comment (ax3l) [inline src/libPMacc/include/math/vector/compile-time/Vector.hpp]: merged :)

**Comment by psychocoderHPC** [comment]
rebased and ready to merge

**Comment by ax3l** [comment]
btw: this is a pretty deep change for users, since many output files change. we need to document that in the changelog (comment added in description)

File: src/picongpu/include/plugins/PluginController.hpp [hunk 6]
```diff
@@ -98,136 +98,112 @@ class PluginController : public ILightweightPlugin
...
-    typedef LiveViewPlugin<PIC_Electrons > LiveImageElectrons;
-    typedef CountParticles<PIC_Electrons> ElectronCounter;
-    typedef EnergyParticles<PIC_Electrons> EnergyElectrons;
-    typedef PositionsParticles<PIC_Electrons> PositionElectrons;
-    typedef BinEnergyParticles<PIC_Electrons> BinEnergyElectrons;
-#if(ENABLE_RADIATION == 1)
-    typedef Radiation<PIC_Electrons> RadiationElectrons;
-#endif
-#endif
+    /** apply vector component one to component zero
```
Comment (ax3l) [inline src/picongpu/include/plugins/PluginController.hpp]: - if it is `vector<dataType,plugin>` than you apply el. 0 in el. 1 :)
-> `apply the 1st vector component to the 2nd`

Comment (ax3l) [inline src/picongpu/include/plugins/PluginController.hpp]: pls update the year in this file when you change it :)

Comment (ax3l) [inline src/picongpu/include/plugins/PluginController.hpp]: I am not sure if you want to add the change commented as a `todo` above with this pull: #818

File: src/picongpu/include/plugins/PluginController.hpp [hunk 7]
```diff
@@ -98,136 +98,112 @@ class PluginController : public ILightweightPlugin
...
-#endif
 
-#if (ENABLE_ADIOS == 1)
-        plugins.push_back(new adios::ADIOSWriter());
+    /* define field plugins */
+    typedef bmpl::vector<
+#if(SIMDIM==DIM3)
+     SliceFieldPrinterMulti<bmpl::_1>
 #endif
+    > AnonymousFieldPlugins;
```
Comment (ax3l) [inline src/picongpu/include/plugins/PluginController.hpp]: `UnspecializedFieldPlugins` ;)

File: src/picongpu/include/plugins/PluginController.hpp [hunk 8]
```diff
@@ -98,136 +98,112 @@ class PluginController : public ILightweightPlugin
...
 
-        plugins.push_back(new EnergyFields("EnergyFields", "energy_fields"));
-        plugins.push_back(new SumCurrents());
+    typedef bmpl::vector< FieldB, FieldE, FieldJ> AllFields;
 
-#if(SIMDIM==DIM3)
+    typedef typename AllCombinations<
+      bmpl::vector<AllFields, AnonymousFieldPlugins>
+    >::type CombinationsAnonymousFieldPlugins;
```
Comment (ax3l) [inline src/picongpu/include/plugins/PluginController.hpp]: `CombinedUnspecializedFieldPlugins`

File: src/picongpu/include/plugins/PluginController.hpp [hunk 9]
```diff
@@ -98,136 +98,112 @@ class PluginController : public ILightweightPlugin
...
-#if(PIC_ENABLE_PNG==1)
-        plugins.push_back(new PngImageIons("PngImageIons", "png_i"));
-#endif
-        plugins.push_back(new BinEnergyIons("BinEnergyIons", "bin_i"));
-        plugins.push_back(new IonCounter("IonsCount", "ions_cnt"));
-        plugins.push_back(new EnergyIons("EnergyIons", "energy_i"));
+      , PerSuperCell<bmpl::_1>
+      , PhaseSpaceMulti<particles::shapes::Counter::ChargeAssignment, bmpl::_1>
 #endif
+    > AnonymousSpeciesPlugins;
```
Comment (ax3l) [inline src/picongpu/include/plugins/PluginController.hpp]: `UnspecializedSpeciesPlugins`

File: src/picongpu/include/plugins/PluginController.hpp [hunk 10]
```diff
@@ -98,136 +98,112 @@ class PluginController : public ILightweightPlugin
...
 
-#if(ENABLE_RADIATION == 1)
-        plugins.push_back(new RadiationElectrons("RadiationElectrons", "radiation_e"));
-#endif
+    typedef typename AllCombinations<
+        bmpl::vector<VectorAllSpecies, AnonymousSpeciesPlugins>
+    >::type CombinationsAnonymousSpeciesPlugins;
```
Comment (ax3l) [inline src/picongpu/include/plugins/PluginController.hpp]: `CombinedUnspecializedSpeciesPlugins`

File: src/picongpu/include/plugins/PluginController.hpp (hunk 6 above)
Comment (psychocoderHPC) [inline src/picongpu/include/plugins/PluginController.hpp]: I change the name of `PngPlugin` in a separate pull

**Comment by psychocoderHPC** [comment]
ready to merge

**Comment by ax3l** [comment]
thank you for the fantastic update!

@ComputationalRadiationPhysics/picongpu-developers @ComputationalRadiationPhysics/picongpu-maintainers take care if you are an early adopter and using `dev` for production. this merge changes a many options you use in your `.cfg` files!

**Comment by bussmann** [comment]
Amazing! Wonderful work!

**Pull request merged**